SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Interop/SpatialWorkerFlags/SpatialWorkerFlagsTest.cpp

```diff
@@ -34,61 +38,353 @@ SPATIALWORKERFLAGS_TEST(GIVEN_a_flagUpdate_op_WHEN_adding_a_worker_flag_THEN_fla
+	SpatialWorkerFlags->RegisterFlagUpdatedCallback("test", WorkerFlagDelegate, ESpatialCallbackStyle::InvokeImmidiatelyIfAlreadySet);
+
+	// THEN
+	TestTrue("Delegate Function was not called", SpyObj->GetTimesFlagUpdated() == 0);
+
+	// Add a different test flag
+	Worker_FlagUpdateOp OpAddFlag = CreateWorkerFlagUpdateOp("test2", "10");
+	SpatialWorkerFlags->ApplyWorkerFlagUpdate(OpAddFlag);
+
+	TestTrue("Delegate Function was not called", SpyObj->GetTimesFlagUpdated() == 0);
+
+	return true;
+}
+
+// a flag await
+
+SPATIALWORKERFLAGS_TEST(GIVEN_no_flags_WHEN_awaiting_flag_update_THEN_delegate_is_not_invoked)
+{
+	// GIVEN
+	// No flags
+
+	// WHEN
+	// Await flag update
+	UWorkerFlagsTestSpyObject* SpyObj = NewObject<UWorkerFlagsTestSpyObject>();
+	FOnWorkerFlagUpdatedBP WorkerFlagDelegate;
+	WorkerFlagDelegate.BindDynamic(SpyObj, &UWorkerFlagsTestSpyObject::SetFlagUpdated);
+
+	USpatialWorkerFlags* SpatialWorkerFlags = NewObject<USpatialWorkerFlags>();
+	SpatialWorkerFlags->AwaitFlagUpdated("test", WorkerFlagDelegate);
+
+	// THEN
+	TestTrue("Delegate Function was not called", SpyObj->GetTimesFlagUpdated() == 0);
+
+	return true;
+}
+
+SPATIALWORKERFLAGS_TEST(GIVEN_an_updated_flag_WHEN_awaiting_flag_update_delegate_THEN_delegate_is_invoked)
+{
+	// GIVEN
+	// Add test flag
+	const FString TestFlagName = TEXT("test");
+	USpatialWorkerFlags* SpatialWorkerFlags = NewObject<USpatialWorkerFlags>();
+	Worker_FlagUpdateOp OpAddFlag = CreateWorkerFlagUpdateOp(TCHAR_TO_ANSI(*TestFlagName), "10");
+	SpatialWorkerFlags->ApplyWorkerFlagUpdate(OpAddFlag);
+
+	// WHEN
+	// Await flag update
+	UWorkerFlagsTestSpyObject* SpyObj = NewObject<UWorkerFlagsTestSpyObject>();
+	FOnWorkerFlagUpdatedBP WorkerFlagDelegate;
+	WorkerFlagDelegate.BindDynamic(SpyObj, &UWorkerFlagsTestSpyObject::SetFlagUpdated);
+	SpatialWorkerFlags->AwaitFlagUpdated(TestFlagName, WorkerFlagDelegate);
+
+	// THEN
+	TestTrue("Delegate Function called once", SpyObj->GetTimesFlagUpdated() == 1);
+
+	return true;
+}
+
+SPATIALWORKERFLAGS_TEST(GIVEN_awaiting_a_flag_update_delegate_WHEN_the_worker_flag_updates_THEN_delegate_is_invoked)
+{
+	// GIVEN
+	// Await flag update
+	const FString TestFlagName = TEXT("test");
+	UWorkerFlagsTestSpyObject* SpyObj = NewObject<UWorkerFlagsTestSpyObject>();
+	FOnWorkerFlagUpdatedBP WorkerFlagDelegate;
+	WorkerFlagDelegate.BindDynamic(SpyObj, &UWorkerFlagsTestSpyObject::SetFlagUpdated);
+	USpatialWorkerFlags* SpatialWorkerFlags = NewObject<USpatialWorkerFlags>();
+	SpatialWorkerFlags->AwaitFlagUpdated(TestFlagName, WorkerFlagDelegate);
+
+	// WHEN
+	// Add test flag
+	Worker_FlagUpdateOp OpAddFlag = CreateWorkerFlagUpdateOp(TCHAR_TO_ANSI(*TestFlagName), "10");
+	SpatialWorkerFlags->ApplyWorkerFlagUpdate(OpAddFlag);
+
+	// THEN
+	TestTrue("Delegate Function called once", SpyObj->GetTimesFlagUpdated() == 1);
+
+	return true;
+}
+
+SPATIALWORKERFLAGS_TEST(GIVEN_awaiting_a_flag_update_delegate_WHEN_the_worker_flag_class_is_destroyed_THEN_delegate_is_not_invoked)
```

Comment:
Shouldn't we also have a test for ensuring the await for a delegate is triggered only once?

Reply:
Yeah that is a good point